pkg/controller/cloud/node_controller.go

```diff
@@ -531,11 +537,11 @@ func ensureNodeExistsByProviderID(ctx context.Context, instances cloudprovider.I
 	return instances.InstanceExistsByProviderID(ctx, providerID)
 }
 
-func getNodeAddressesByProviderIDOrName(ctx context.Context, instances cloudprovider.Instances, node *v1.Node) ([]v1.NodeAddress, error) {
-	nodeAddresses, err := instances.NodeAddressesByProviderID(ctx, node.Spec.ProviderID)
+func getNodeAddressesByProviderIDOrName(ctx context.Context, instances cloudprovider.Instances, providerID, nodeName string) ([]v1.NodeAddress, error) {
+	nodeAddresses, err := instances.NodeAddressesByProviderID(ctx, providerID)
 	if err != nil {
 		providerIDErr := err
-		nodeAddresses, err = instances.NodeAddresses(ctx, types.NodeName(node.Name))
+		nodeAddresses, err = instances.NodeAddresses(ctx, types.NodeName(nodeName))
```

Comment:
This isn't going to work for GCP using the existing GCE cloud provider implementation. (https://github.com/kubernetes/legacy-cloud-providers/blob/master/gce/gce_instances.go#L86) The gce implementation has the  issue detailed in the todo of the interface (https://github.com/kubernetes/cloud-provider/blob/master/cloud.go#L160). It was written back when this was only called from the Kubelet and is written such that it will only do the right thing, when called from the Kubelet.

Reply:
Agreed and I think this is the proper fix. `node.Spec.ProviderID` being empty is acceptable which is why we have checks for getting by name.

Reply:
Yes, but this is not something related to this PR. That's why I propose this PR to fix the `providerID` passing to `getNodeAddressesByProviderIDOrName` and in this way prevent the fallback call to `NodeAddresses` which fails for gce when the cloud-controller-manager is not running on the host.

Reply:
Agreed, which is why I lgtm'd it. I just wanted to make sure you were aware of the issue. 

The Kubelet usage of the cloud provider implementation should be going away soon. We should fix the GCE implementation to do the right thing in non Kubelet use cases.

Reply:
> Yes, but this is not something related to this PR. 

This is the part I'm a bit confused about. I think it is related because implementing the GCP fallback makes this a non-issue. We should treat `node.Spec.ProviderID` as the source of truth because we don't know yet if the provider ID from `GetInstanceProviderID` is valid. There are also some cases where `node.Spec.ProviderID` can be set from another source and `GetInstanceProviderID` returns `cloudprovider.NotImplemented`.

Reply:
> This is the part I'm a bit confused about. I think it is related because implementing the GCP fallback makes this a non-issue. 

I don't agree here. Currently all `getNodeAddressesByProviderIDOrName`, `getInstanceTypeByProviderIDOrName`, `getZoneByProviderIDOrName` are called with `node.Spec.ProviderID=""` which makes pointless all of the provideID usage in these funcs such as `instances.NodeAddressesByProviderID(ctx, node.Spec.ProviderID)`, `instances.InstanceTypeByProviderID(ctx, providerID)`, `zones.GetZoneByProviderID(ctx, providerID)`.

Generally I agree that the GCE provider handling for `NodeAddresses` func can be improved to do not rely that the cloud-controller-manager is running on the Node. But this is something different.

Reply:
> I don't agree here. Currently all getNodeAddressesByProviderIDOrName, getInstanceTypeByProviderIDOrName, getZoneByProviderIDOrName are called with node.Spec.ProviderID="" which makes pointless all of the provideID usage in these funcs such as instances.NodeAddressesByProviderID(ctx, node.Spec.ProviderID), instances.InstanceTypeByProviderID(ctx, node.Spec.ProviderID), zones.GetZoneByProviderID(ctx, node.Spec.ProviderID).

The ProviderID can be preset if kubelet has `--provider-id` set, but I see this doesn't break that since `node.Spec.ProviderID` takes precedent.

Reply:
https://github.com/kubernetes/kubernetes/issues/90185